src/details/ast.hpp

```diff
@@ -691,62 +712,111 @@ class Table {
      */
     auto fields() const -> std::vector<Field>;
     auto range() const -> minilua::Range;
-    auto raw() const -> ts::Node;
+    auto debug_print() const -> std::string;
 };
 // a few empty classes that are just used as additional return types
 class Spread {};
 class Self {};
 class Break {};
+using PrefixVariant = std::variant<Self, VariableDeclarator, FunctionCall, Expression>;
 /**
  * class for prefix nodes
  */
 class Prefix {
-    ts::Node prefix;
+    using modifiedPrefixVariant =
+        std::variant<Self, VariableDeclarator, FunctionCall, std::shared_ptr<Expression>>;
+    using PrefixTuple = std::tuple<modifiedPrefixVariant, minilua::Range, GEN_CAUSE>;
+    enum TupleIndex { PFX_VARIANT, RANGE, CAUSE };
+    std::variant<ts::Node, PrefixTuple> content;
 
 public:
     explicit Prefix(ts::Node);
+    explicit Prefix(const VariableDeclarator&, GEN_CAUSE);
+    explicit Prefix(const FunctionCall&, GEN_CAUSE);
     /**
      *
      * @return a variant containing the class this Prefix gets resolved to
      */
-    auto options() const -> std::variant<Self, VariableDeclarator, FunctionCall, Expression>;
+    auto options() const -> PrefixVariant;
     auto range() const -> minilua::Range;
-    auto raw() const -> ts::Node;
+    auto debug_print() const -> std::string;
+    auto to_string() const -> std::string;
 };
+using ExpressionVariant = std::variant<
+    Spread, Prefix, FunctionDefinition, Table, BinaryOperation, UnaryOperation, Literal,
+    Identifier>;
 /**
  * class for expression nodes
  */
 class Expression {
-    ts::Node exp;
+    using ExpTup = std::tuple<ExpressionVariant, minilua::Range>;
+    std::variant<ts::Node, std::tuple<ExpressionVariant, minilua::Range>> content;
+    enum TupleIndex { EXP_VARIANT, RANGE };
 
 public:
     explicit Expression(ts::Node);
+    explicit Expression(const BinaryOperation&);
+    explicit Expression(const UnaryOperation&);
+    explicit Expression(const FunctionDefinition&);
+    explicit Expression(const Prefix&);
+    explicit Expression(const Literal&);
+    explicit Expression(const Identifier&);
```

Comment:
Is it useful to take `const&` if you store values? This will just copy everything, right?